deltachat-jsonrpc/src/api/mod.rs

```diff
@@ -1352,6 +1352,57 @@ impl CommandApi {
         .await
     }
 
+    /// Offers a backup for remote devices to retrieve.
+    ///
+    /// Can be cancelled by stopping the ongoing process.  Success or failure can be tracked
+    /// via the `ImexProgress` event which should either reach `1000` for success or `0` for
+    /// failure.
+    ///
+    /// This **stops IO** while it is running.
+    ///
+    /// Returns once a remote device has retrieved the backup.
+    async fn provide_backup(&self, account_id: u32) -> Result<()> {
+        let ctx = self.get_context(account_id).await?;
+        ctx.stop_io().await;
+        let provider = match imex::BackupProvider::prepare(&ctx).await {
+            Ok(provider) => provider,
+            Err(err) => {
+                ctx.start_io().await;
```

Comment:
Interesting idea, but we can't run async code in `Drop` unless we spawn it.  However we really want to be rather sure that IO is running when we return from this function, otherwise we'll create race conditions.  So I don't think we can do this, as nice as it would be.

Reply:
I'm not sure what you have in mind.  Are you thinking of something like `with_stopped_io<F: FnOnce(F: FnOnce() -> dyn Future<Output = ()>>() { todo!() }`?

I don't really think I want to do this in the current PR either.  I think this is again a case of following the current way of doing things and the function is small and clear enough right now.  I also don't want to mix this in with improving the general infrastructure, I think those are best done in separate PRs.  But sure, doing those things are worthwhile!

Reply:
I'm not sure we need such a generic approach (but if we have several places where IO must be temporarily stopped, it's a good idea). I meant we can add `async fn provide_backup_inner(&self, account_id: u32) -> Result<()>` which is called by `provide_backup()` doing `stop_io()` and `start_io()` so that the latter can't be forgotten when handling new erorrs f.e.

Reply:
I'm going to leave this for now, it's a bit of a judgement call but I don't think handling it this way here is much of an issue for now and I'm not convinced it will become much more readable or maintainable with such a change.  If the complexity of this function increases my opinion might change though.

Reply:
It's not critical, yes, but as i came from C, for me it's strange that in Rust it looks worse :) because in C you can `goto out;` / do a fake loop (if you don't like `goto`) / use single-return-point approach. The result is having all necessary cleanup/rollback in the only place. But in Rust i don't really understand what is the proper way of doing that